ehrql/measures/calculate.py

```diff
@@ -0,0 +1,177 @@
+                totals = accumulator[group]
+                totals[0] += numerator
+                totals[1] += denominator
+
+        for measure, (_, accumulator) in zip(self.measures, fetcher_accumulator_pairs):
+            for group, (numerator, denominator) in accumulator.items():
+                yield measure, numerator, denominator, group
+
+    def add_measure(self, measure):
+        # Record denominator and intervals from first measure
+        if self.denominator is None:
+            self.denominator = measure.denominator
+            self.intervals = measure.intervals
+            self.variables["population"] = series_as_bool(self.denominator)
+        else:
+            assert measure.denominator == self.denominator
+            assert measure.intervals == self.intervals
+
+        numerator_index = self.add_variable(series_as_int(measure.numerator))
+        denominator_index = self.add_variable(series_as_int(measure.denominator))
+        group_indexes = [
+            self.add_variable(column) for column in measure.group_by.values()
+        ]
+
+        # Create a function which takes a results row and returns a numerator,
+        # denominator, and tuple of group values, based on their indices
+        fetcher = self.create_fetcher(numerator_index, denominator_index, group_indexes)
+
+        self.measures.append(measure)
+        self.fetchers.append(fetcher)
+
+    def add_variable(self, variable):
+        # Return the position of `variable` in the variables dict, adding it if not
+        # already present
+        try:
+            return list(self.variables.values()).index(variable)
+        except ValueError:
+            index = len(self.variables)
+            self.variables[f"column_{index}"] = variable
+            return index
+
+    @staticmethod
+    def create_fetcher(numerator_index, denominator_index, group_indexes):
+        # Given a bunch of indices we want a function which extracts just those indices
+        # from a tuple. This is going to be called frequently, so the fastest way to do
+        # this is to build a function definition and then eval it.
+        group_items = [f"row[{i}]" for i in group_indexes]
+        items = [
+            f"row[{numerator_index}]",
+            f"row[{denominator_index}]",
+            f"({', '.join(group_items)},)",
+        ]
+        code = f"lambda row: ({', '.join(items)})"
+        return eval(code)
+
+
+def series_as_bool(series):
+    series_type = get_series_type(series)
+    if series_type is bool:
+        return series
+    elif series_type is int:
+        return Function.GT(series, Value(0))
+    else:
+        assert False
+
+
+def series_as_int(series):
+    series_type = get_series_type(series)
+    if series_type is int:
+        return series
+    elif series_type is bool:
+        # TODO: This is definitely not the most efficient way to do this. We should
+        # extend the `CastToInt` operation to apply to boolean as well.
```

Comment:
Can you add a ticket for this so we don't lose sight of it?

Reply:
Done:
https://github.com/opensafely-core/ehrql/issues/1313